posts/lint-rcpp/index.qmd

```diff
@@ -0,0 +1,168 @@
+---
+title: "Improving the C++ Code Quality of an Rcpp Package"
+author: "Pratik R Gupte"
+date: "2022-12-02"
+categories: [code quality, R package, Rcpp]
+format:
+  html:
+    toc: true
+---
+
+## Use case
+
+The R package development ecosystem includes packages such as [{lintr}](https://lintr.r-lib.org/) and [{styler}](https://styler.r-lib.org) that can help to check code style, and to format R code.
+
+However, these packages cannot lint or style the C++ code of [{Rcpp}](https://www.rcpp.org/) packages. This could leave the C++ code of an Rcpp package less clean than the R code, increasing the technical debt already associated with using two-languages.
+
+In the Epiverse, we encounter this issue with [{finalsize}](https://epiverse-trace.github.io/finalsize/), and we anticipate the same issue with further epidemic modelling packages that we seek to develop or adapt, such as [{fluEvidenceSynthesis}](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005838).
+
+## Choice of C++ linters
+
+C++ linters such as [_clang-tidy_](https://clang.llvm.org/extra/clang-tidy/) stumble when dealing with C++ code in `src/`, as the _clang_ toolchain attempts to compile it. This does not work for Rcpp packages, as the `Rcpp.h` header cannot be found --- this linking is handled by {Rcpp}.
+
+Fortunately, other C++ linters and code checking tools are available and can be used safely with Rcpp packages.
+
+We have chosen to use _cpplint_ and _cppcheck_ for {finalsize}.
+
+### Cpplint
+
+[_cpplint_](https://github.com/cpplint/cpplint) is a tool that checks whether C/C++ files follow [Google's C++ style guide](http://google.github.io/styleguide/cppguide.html). _cpplint_ is (1) Easy to install across platforms, (2) Easily instructed to ignore the `RcppExports.cpp` file, so as to preserve its formatting, and (3) Does not error when it cannot find `Rcpp.h`.
+
+_cpplint_ can be run locally from the terminal, and instructed to not lint `RcppExports.cpp`, but all other `.cpp` files in `src/`.
+
+```sh
+cpplint --exclude="src/RcppExports.cpp" src/*.cpp
+```
+
+### Cppcheck
+
+[_cppcheck_](https://cppcheck.sourceforge.io/) is a static code analysis tool, that aims to "have very few false positives". This is especially useful for the non-standard organisation of Rcpp projects compared to C++ projects.
+
+_cppcheck_ can also be run locally and instructed to ignore the autogenerated `RcppExports.cpp` file, while throwing up issues with style.
+
+```sh
+cppcheck -i src/RcppExports.cpp --enable=style --error-exitcode=1 src
+```
+
+Here, the `--enable=style` option lets _cppcheck_ flag issues with style, acting as a second linter.
```

Comment:
Is it possible to enable more checks and selectively remove only those that are not compatible with Rcpp? You mention `missingIncludeSystem` and `unusedFunction` below as incompatible but presumably by setting `--enable=style`, we're excluding more than what we should and some checks may work with Rcpp packages.

Reply:
The ones that are being excluded (`missingIncludeSystem`; keyword `missingInclude`, and `unusedFunction`) are the only other un-enabled options. Setting `--enable=style` enables `style`, `performance`, and `portability` as well. I haven't noticed a difference between this and `--enable=warning`, so I have left it at `--enable=warning` for now.

The _cppcheck_ documentation is not great, but this Debian documentation has the same details: https://manpages.debian.org/testing/cppcheck/cppcheck.1.en.html

Reply:
Again, I'm not sure what the correct option is - I think `warning` is a reasonable choice which also flags `style` issues.